lnwallet/channel.go

```diff
@@ -3470,44 +3425,37 @@ func newHtlcResolution(signer Signer, localChanCfg *channeldb.ChannelConfig,
 	// transaction creates so we can generate the signDesc required to
 	// complete the claim process after a delay period.
 	htlcSweepScript, err := secondLevelHtlcScript(
-		revokeKey, delayKey, csvDelay,
+		keys.revocationKey, keys.delayKey, csvDelay,
 	)
 	if err != nil {
 		return nil, err
 	}
 
-	return &OutgoingHtlcResolution{
+	// TODO: Signing with the delay key is wrong for remote commitments
```

Comment:
I don't follow? Signing with the delay key is the proper key to use for remote commitments. Or perhaps you're referecnign a bug which was recently fixed by this (30d5b9062617d003b1f6c83f3c576a8efc57b0c1) commit?

Reply:
That fixes one issue, but I'm pretty sure there's another bug still. Correct me if I'm wrong.

If the other person publishes a commitment transaction (the remote commitment), we want to redeem all HTLC outputs spendable by us. There are two cases where an HTLC output on the remote commitment is spendable by us: 1) it is an offered HTLC output (them to us) and we have the payment preimage or 2) it is a received HTLC output (us to them) that has expired. *In neither case do we sign with our delay key, we sign with our local key, and provide the payment preimage in the first case.* You can follow the code from here: https://github.com/lightningnetwork/lnd/blob/master/lnwallet/channel.go#L1404.

Reply:
I'm familiar with the code. 

In the second case we _do_ sign the the delay key as we still incur the relative CSV delay to sweep the output after the second-level HTLC transaction has confirmed. In the first case we will still _eventually_ sign with the delay key as we need to incur the CSV delay from the HTLC success transaction. 

As a result, both the local key _and_ the delay key need their tweaks populated within the key ring.

Reply:
The current code doesn't yet handle the HTLC success case (receiving an HTLC and we need to go to chain in order to redeem it). I have a branch locally which I'll be putting up once we get the preliminary code to handle the second-level timeout transactions in the `utoxNursery`.